frontend/catalyst/jax_primitives.py

```diff
@@ -1657,29 +1657,41 @@ def _hamiltonian_lowering(jax_ctx: mlir.LoweringRuleContext, coeffs: ir.Value, *
 def _sample_abstract_eval(obs, shots, shape):
     assert isinstance(obs, AbstractObs)
 
-    if obs.primitive is compbasis_p:
-        assert shape == (shots, obs.num_qubits)
-    else:
-        assert shape == (shots,)
+    if isinstance(shots, int):
+        if obs.primitive is compbasis_p:
+            assert shape == (shots, obs.num_qubits)
+        else:
+            assert shape == (shots,)
 
-    return core.ShapedArray(shape, jax.numpy.float64)
+    return core.DShapedArray(shape, jax.numpy.dtype("float64"))
 
 
 @sample_p.def_impl
 def _sample_def_impl(ctx, obs, shots, shape):  # pragma: no cover
     raise NotImplementedError()
 
 
-def _sample_lowering(jax_ctx: mlir.LoweringRuleContext, obs: ir.Value, shots: int, shape: tuple):
+def _sample_lowering(
+    jax_ctx: mlir.LoweringRuleContext, obs: ir.Value, shots: Union[int, ir.Value], shape: tuple
+):
+    # Note: result shape of sample op is (shots, number_of_qubits)
     ctx = jax_ctx.module_context.context
     ctx.allow_unregistered_dialects = True
 
-    i64_type = ir.IntegerType.get_signless(64, ctx)
-    shots_attr = ir.IntegerAttr.get(i64_type, shots)
     f64_type = ir.F64Type.get()
-    result_type = ir.RankedTensorType.get(shape, f64_type)
+    i64_type = ir.IntegerType.get_signless(64, ctx)
 
-    return SampleOp(result_type, obs, shots_attr).results
+    if isinstance(shots, int):
+        result_type = ir.RankedTensorType.get(shape, f64_type)
+        return SampleOp(result_type, obs, static_shots=shots).results
+
+    else:
+        # TODO: use actual dynamic shape tensor type API
+        # Somehow this magic number means dynamically sized dimensions in jax
+        result_shape = (-9223372036854775808, shape[1])
```

Comment:
Shouldn't we store it in some kind of named constant, so we know where it comes from?

Reply:
It is now no longer used, and we use the proper `ir.ShapedType.get_dynamic_size()` instead (which just returns this number).